hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -402,35 +438,113 @@ public void addConnectionHeartbeatListener(ConnectionHeartbeatListener connectio
+                            connection.setIsAuthenticatedAsOwner();
+                            clusterService.setPrincipal(new ClientPrincipal(result.uuid, result.ownerUuid));
+                        }
+                        authenticated(target, connection);
+                        callback.onSuccess(connection);
+                        break;
+                    case CREDENTIALS_FAILED:
+                        AuthenticationException e = new AuthenticationException("Invalid credentials!");
+                        failed(target, connection, e);
+                        callback.onFailure(e);
+                        break;
+                    default:
+                        AuthenticationException exception =
+                                new AuthenticationException("Authentication status code not supported. status:"
+                                        + authenticationStatus);
+                        failed(target, connection, exception);
+                        callback.onFailure(exception);
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                failed(target, connection, t);
+                callback.onFailure(t);
+            }
+        }, executionService.getInternalExecutor());
+    }
+
     private class InitConnectionTask implements Runnable {
 
         private final Address target;
-        private final Address remoteAddress;
-        private final Authenticator authenticator;
+        private final boolean asOwner;
+        private final AuthenticationCallback callback;
 
-        InitConnectionTask(Address target, Address remoteAddress, Authenticator authenticator) {
+        InitConnectionTask(Address target, boolean asOwner, AuthenticationCallback callback) {
             this.target = target;
-            this.remoteAddress = remoteAddress;
-            this.authenticator = authenticator;
+            this.asOwner = asOwner;
+            this.callback = callback;
         }
 
         @Override
         public void run() {
-            final Object lock = getLock(target);
-            synchronized (lock) {
-                ClientConnection connection = connections.get(target);
-                if (connection != null) {
-                    return;
-                }
+            ClientConnection connection = connections.get(target);
+            if (connection == null) {
                 try {
-                    initializeConnection(remoteAddress, authenticator);
-                } catch (Exception e) {
+                    connection = createSocketConnection(target);
+                } catch (IOException e) {
                     logger.finest(e);
-                } finally {
+                    callback.onFailure(e);
                     connectionsInProgress.remove(target);
+                    return;
                 }
-
             }
+
+            authenticate(target, connection, asOwner, callback);
```

Comment:
What if `authenticate()` itself fails with an exception? You should remove `target` from `connectionsInProgress` when `authenticate()` fails.